src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -4025,7 +4029,7 @@ let rec universe_of_aux env e =
         | Tm_meta _
         | Tm_type _ ->
           universe_of_aux env hd, args
-        | Tm_match(_, _, hd::_) ->  //AR: TODO: use return annotation?
+        | Tm_match(_, _, hd::_, _) ->  //AR: TODO: use return annotation? Or the residual_comp?
```

Comment:
I guess this is the main question I have on this PR. I mentioned it when we spoke, I think. Which is, is there some redundancy between this additional residual_comp and the return annotation? Why not always populate the return annotation after typechecking a match and then using that annotation to drive the fast path? We could keep a flag in the return annotation to indicate if it was user-annotated or if it was inferred.